File: BlogApp-RN/src/features/authentication/domain/usecases/signUp.ts
import { AsyncResult } from "../../../../core/typedefs/asyncResult";
import { AuthResponse } from "../../data/models/authResponse";
import AuthRepo from "../repos/authRepo";

class SignUp {
  constructor(private repo: AuthRepo) {}

  call: (username: string, password: string) => AsyncResult<AuthResponse> = (
    username,
    password
  ) => {
    return this.repo.signUp(username, password);
  };
}

export default SignUp;

File: blogapp_flutter/lib/features/authentication/domain/usecases/sign_up.dart
import 'package:blogapp_flutter/core/typedefs/usecases.dart';

import '../../../../core/typedefs/async_result.dart';
import '../../data/models/auth_response.dart';
import '../repos/auth_repo.dart';

class SignUp extends UsecaseWithParams<AuthResponse, SignUpParams> {
  final AuthRepo _repo;

  SignUp(this._repo);

  @override
  AsyncResult<AuthResponse> call(params) {
    return _repo.signUp(username: params.username, password: params.password);
  }
}

class SignUpParams {
  final String username;
  final String password;
  SignUpParams(this.username, this.password);
}
